File: spiralTraverse.java
/*
In this problem we need to travers 
CASE 1. LEFT --> RIGHT
CASE 2. UP --> DOWN
CASE 3. RIGHT --> LEFT
CASE 4. BOTTOM --> UP

4 directions recursive way

TC --> O(n*m)
SC --> O(1)
 */
import java.util.*;

class Solution {

    public List<Integer> spiralOrder(int[][] matrix) {
        int left = 0;
        int right = matrix[0].length - 1;
        int top = 0;
        int bottom = matrix.length - 1;
        List<Integer> res = new ArrayList<>();

        while (top <= bottom && left <= right) {
            //case 1 left to right
            for (int i = left; i <= right; i++) {
                res.add(matrix[top][i]);

            }//for
            top++;

            //case 2 top to bottom
            for (int j = top; j <= bottom; j++) {
                res.add(matrix[j][right]);

            }//for

            right--;

            //case 3 right to left
            if (top <= bottom) {
                for (int i = right; i >= left; i--) {
                    res.add(matrix[bottom][i]);

                }//for

                bottom--;
            }//if

            //case 4 bottom to top
            if (left <= right) {

                for (int j = bottom; j >= top; j--) {
                    res.add(matrix[j][left]);

                }
                left++;
            }

        }//while

        return res;

    }//method
}//class

File: spiral_traverse.cpp
// Time : o(mn)
// space : o(1) since output is not counted in space complexity

class Solution {
public:
    vector<int> spiralOrder(vector<vector<int>>& matrix) {
        int dir = 1; // right = 1, down = 2, left = 3, up = 4
        vector<int> answer(matrix.size()*matrix[0].size());


        int right_edge = matrix[0].size() - 1; // right most column 
        int down_edge = matrix.size() -1; // bottom most row
        int left_edge = 0; // left most column
        int up_edge = 0; // top most row

        int r = 0; 
        int c = 0;
        for(int i = 0; i < answer.size(); i++)
        {
            answer[i] = matrix[r][c];
            if(dir == 1) // right
            {
                if(c == right_edge)
                {
                    // change direction to down
                    dir = 2;
                    r++;
                    // change edge for upward traversal
                    up_edge++;
                }
                else
                {
                    c++;
                }

            }
            else if(dir == 2) // down
            {
                if(r == down_edge)
                {
                    // change direction to left
                    dir = 3;
                    c--;
                    // change edge for right traversal
                    right_edge--;
                }
                else
                {
                    r++;
                }

            }
            else if(dir == 3) // left
            {
                if(c == left_edge)
                {
                    // change direction to up
                    dir = 4;
                    r--;
                    // change edge for down traversal
                    down_edge--;
                }
                else
                {
                    c--;
                }
            }
            else // up
            {
                if(r == up_edge)
                {
                    // change direction to right
                    dir = 1;
                    c++;
                    // change edge for left traversal
                    left_edge++;
                }
                else
                {
                    r--;
                }

            }
        }
    return answer;
    }
    
};